default thread timeout too short for CI

Since this morning [our CI](https://github.com/openpathsampling/openpathsampling/runs/3563720388?) started failing with:

```
Run python devtools/autorelease_check.py
Traceback (most recent call last):
  File "devtools/autorelease_check.py", line 24, in <module>
    repo_path='.'
  File "/usr/share/miniconda/envs/test/lib/python3.7/site-packages/autorelease/check_runners.py", line 52, in __init__
    repo_path=repo_path
  File "/usr/share/miniconda/envs/test/lib/python3.7/site-packages/autorelease/git_repo_checks.py", line 28, in __init__
    self.repo.remotes.origin.fetch()
  File "/usr/share/miniconda/envs/test/lib/python3.7/site-packages/git/remote.py", line 856, in fetch
    res = self._get_fetch_info_from_stderr(proc, progress)
  File "/usr/share/miniconda/envs/test/lib/python3.7/site-packages/git/remote.py", line 727, in _get_fetch_info_from_stderr
    handle_process_output(proc, None, progress_handler, finalizer=None, decode_streams=False)
  File "/usr/share/miniconda/envs/test/lib/python3.7/site-packages/git/cmd.py", line 151, in handle_process_output
    raise RuntimeError(f"Thread join() timed out in cmd.handle_process_output(). Timeout={timeout} seconds")
RuntimeError: Thread join() timed out in cmd.handle_process_output(). Timeout=10.0 seconds
Fatal Python error: could not acquire lock for <_io.BufferedReader name=4> at interpreter shutdown, possibly due to daemon threads

Thread 0x00007fbb2f2d7700 (most recent call first):
  File "/usr/share/miniconda/envs/test/lib/python3.7/site-packages/git/cmd.py", line 103 in pump_stream
  File "/usr/share/miniconda/envs/test/lib/python3.7/threading.py", line 870 in run
  File "/usr/share/miniconda/envs/test/lib/python3.7/threading.py", line 926 in _bootstrap_inner
  File "/usr/share/miniconda/envs/test/lib/python3.7/threading.py", line 890 in _bootstrap

Current thread 0x00007fbb3332d180 (most recent call first):
  File "/usr/share/miniconda/envs/test/lib/python3.7/site-packages/git/cmd.py", line 406 in __del__
/home/runner/work/_temp/ff824c06-ff56-4ee4-9e2a-e65258688cf4.sh: line 1:  1922 Aborted                 (core dumped) python devtools/autorelease_check.py
Error: Process completed with exit code 134.
```

This seems a side effect of the timeout introduced by #1318 

Could this time-out default to `60 s` instead (or be taken and propagated by the `fetch` function)?


I could make a PR if needed, with either of the two solutions

--- a/git/cmd.py
+++ b/git/cmd.py
@@ -79,7 +79,7 @@ def handle_process_output(process: 'Git.AutoInterrupt' | Popen,
                           finalizer: Union[None,
                                            Callable[[Union[subprocess.Popen, 'Git.AutoInterrupt']], None]] = None,
                           decode_streams: bool = True,
-                          timeout: Union[None, float] = None) -> None:
+                          kill_after_timeout: Union[None, float] = None) -> None:
     """Registers for notifications to learn that process output is ready to read, and dispatches lines to
     the respective line handlers.
     This function returns once the finalizer returns
@@ -94,7 +94,10 @@ def handle_process_output(process: 'Git.AutoInterrupt' | Popen,
         their contents to handlers.
         Set it to False if `universal_newline == True` (then streams are in text-mode)
         or if decoding must happen later (i.e. for Diffs).
-    :param timeout: float, or None timeout to pass to t.join() in case it hangs. Default = None.
+    :param kill_after_timeout:
+        float or None, Default = None
+        To specify a timeout in seconds for the git command, after which the process
+        should be killed.
     """
     # Use 2 "pump" threads and wait for both to finish.
     def pump_stream(cmdline: List[str], name: str, stream: Union[BinaryIO, TextIO], is_decode: bool,
@@ -108,9 +111,12 @@ def handle_process_output(process: 'Git.AutoInterrupt' | Popen,
                         handler(line_str)
                     else:
                         handler(line)
+
         except Exception as ex:
             log.error(f"Pumping {name!r} of cmd({remove_password_if_present(cmdline)}) failed due to: {ex!r}")
-            raise CommandError([f'<{name}-pump>'] + remove_password_if_present(cmdline), ex) from ex
+            if "I/O operation on closed file" not in str(ex):
+                # Only reraise if the error was not due to the stream closing
+                raise CommandError([f'<{name}-pump>'] + remove_password_if_present(cmdline), ex) from ex
         finally:
             stream.close()
 
@@ -146,9 +152,16 @@ def handle_process_output(process: 'Git.AutoInterrupt' | Popen,
     ## FIXME: Why Join??  Will block if `stdin` needs feeding...
     #
     for t in threads:
-        t.join(timeout=timeout)
+        t.join(timeout=kill_after_timeout)
         if t.is_alive():
-            raise RuntimeError(f"Thread join() timed out in cmd.handle_process_output(). Timeout={timeout} seconds")
+            if hasattr(process, 'proc'):  # Assume it is a Git.AutoInterrupt:
+                process._terminate()
+            else:  # Don't want to deal with the other case
+                raise RuntimeError(f"Thread join() timed out in cmd.handle_process_output()."
+                                   " kill_after_timeout={kill_after_timeout} seconds")
+            if stderr_handler:
+                stderr_handler("error: process killed because it timed out."
+                               f" kill_after_timeout={kill_after_timeout} seconds")
 
     if finalizer:
         return finalizer(process)
@@ -386,13 +399,15 @@ class Git(LazyMixin):
         The wait method was overridden to perform automatic status code checking
         and possibly raise."""
 
-        __slots__ = ("proc", "args")
+        __slots__ = ("proc", "args", "status")
 
         def __init__(self, proc: Union[None, subprocess.Popen], args: Any) -> None:
             self.proc = proc
             self.args = args
+            self.status = None
 
-        def __del__(self) -> None:
+        def _terminate(self) -> None:
+            """Terminate the underlying process"""
             if self.proc is None:
                 return
 
@@ -408,6 +423,7 @@ class Git(LazyMixin):
             # did the process finish already so we have a return code ?
             try:
                 if proc.poll() is not None:
+                    self.status = proc.poll()
                     return None
             except OSError as ex:
                 log.info("Ignored error after process had died: %r", ex)
@@ -419,7 +435,7 @@ class Git(LazyMixin):
             # try to kill it
             try:
                 proc.terminate()
-                proc.wait()    # ensure process goes away
+                self.status = proc.wait()    # ensure process goes away
             except OSError as ex:
                 log.info("Ignored error after process had died: %r", ex)
             except AttributeError:
@@ -430,6 +446,11 @@ class Git(LazyMixin):
                 if is_win:
                     call(("TASKKILL /F /T /PID %s 2>nul 1>nul" % str(proc.pid)), shell=True)
             # END exception handling
+
+
+
+        def __del__(self) -> None:
+            self._terminate()
 
         def __getattr__(self, attr: str) -> Any:
             return getattr(self.proc, attr)
@@ -447,21 +468,26 @@ class Git(LazyMixin):
 
             if self.proc is not None:
                 status = self.proc.wait()
+                p_stderr = self.proc.stderr
+            else: #Assume the underlying proc was killed earlier or never existed
+                status = self.status
+                p_stderr = None
 
-                def read_all_from_possibly_closed_stream(stream: Union[IO[bytes], None]) -> bytes:
-                    if stream:
-                        try:
-                            return stderr_b + force_bytes(stream.read())
-                        except ValueError:
-                            return stderr_b or b''
-                    else:
+            def read_all_from_possibly_closed_stream(stream: Union[IO[bytes], None]) -> bytes:
+                if stream:
+                    try:
+                        return stderr_b + force_bytes(stream.read())
+                    except ValueError:
                         return stderr_b or b''
+                else:
+                    return stderr_b or b''
 
-                if status != 0:
-                    errstr = read_all_from_possibly_closed_stream(self.proc.stderr)
-                    log.debug('AutoInterrupt wait stderr: %r' % (errstr,))
-                    raise GitCommandError(remove_password_if_present(self.args), status, errstr)
             # END status handling
+
+            if status != 0:
+                errstr = read_all_from_possibly_closed_stream(p_stderr)
+                log.debug('AutoInterrupt wait stderr: %r' % (errstr,))
+                raise GitCommandError(remove_password_if_present(self.args), status, errstr)
             return status
 
     # END auto interrupt
@@ -694,7 +720,7 @@ class Git(LazyMixin):
                 as_process: bool = False,
                 output_stream: Union[None, BinaryIO] = None,
                 stdout_as_string: bool = True,
-                kill_after_timeout: Union[None, int] = None,
+                kill_after_timeout: Union[None, float] = None,
                 with_stdout: bool = True,
                 universal_newlines: bool = False,
                 shell: Union[None, bool] = None,

--- a/git/cmd.py
+++ b/git/cmd.py
@@ -79,7 +79,7 @@ def handle_process_output(process: 'Git.AutoInterrupt' | Popen,
                           finalizer: Union[None,
                                            Callable[[Union[subprocess.Popen, 'Git.AutoInterrupt']], None]] = None,
                           decode_streams: bool = True,
-                          timeout: float = 60.0) -> None:
+                          timeout: Union[None, float] = None) -> None:
     """Registers for notifications to learn that process output is ready to read, and dispatches lines to
     the respective line handlers.
     This function returns once the finalizer returns

--- a/git/cmd.py
+++ b/git/cmd.py
@@ -409,6 +409,10 @@ class Git(LazyMixin):
 
         __slots__ = ("proc", "args", "status")
 
+        # If this is non-zero it will override any status code during
+        # _terminate, used to prevent race conditions in testing
+        _status_code_if_terminate: int = 0
+
         def __init__(self, proc: Union[None, subprocess.Popen], args: Any) -> None:
             self.proc = proc
             self.args = args
@@ -427,11 +431,10 @@ class Git(LazyMixin):
                 proc.stdout.close()
             if proc.stderr:
                 proc.stderr.close()
-
             # did the process finish already so we have a return code ?
             try:
                 if proc.poll() is not None:
-                    self.status = proc.poll()
+                    self.status = self._status_code_if_terminate or proc.poll()
                     return None
             except OSError as ex:
                 log.info("Ignored error after process had died: %r", ex)
@@ -443,7 +446,9 @@ class Git(LazyMixin):
             # try to kill it
             try:
                 proc.terminate()
-                self.status = proc.wait()    # ensure process goes away
+                status = proc.wait()  # ensure process goes away
+
+                self.status = self._status_code_if_terminate or status
             except OSError as ex:
                 log.info("Ignored error after process had died: %r", ex)
             except AttributeError:
@@ -849,7 +854,7 @@ class Git(LazyMixin):
 
         if is_win:
             cmd_not_found_exception = OSError
-            if kill_after_timeout:
+            if kill_after_timeout is not None:
                 raise GitCommandError(redacted_command, '"kill_after_timeout" feature is not supported on Windows.')
         else:
             cmd_not_found_exception = FileNotFoundError  # NOQA # exists, flake8 unknown @UndefinedVariable
@@ -916,7 +921,7 @@ class Git(LazyMixin):
             return
         # end
 
-        if kill_after_timeout:
+        if kill_after_timeout is not None:
             kill_check = threading.Event()
             watchdog = threading.Timer(kill_after_timeout, _kill_process, args=(proc.pid,))
 
@@ -927,10 +932,10 @@ class Git(LazyMixin):
         newline = "\n" if universal_newlines else b"\n"
         try:
             if output_stream is None:
-                if kill_after_timeout:
+                if kill_after_timeout is not None:
                     watchdog.start()
                 stdout_value, stderr_value = proc.communicate()
-                if kill_after_timeout:
+                if kill_after_timeout is not None:
                     watchdog.cancel()
                     if kill_check.is_set():
                         stderr_value = ('Timeout: the command "%s" did not complete in %d '

--- a/git/cmd.py
+++ b/git/cmd.py
@@ -154,14 +154,22 @@ def handle_process_output(process: 'Git.AutoInterrupt' | Popen,
     for t in threads:
         t.join(timeout=kill_after_timeout)
         if t.is_alive():
-            if hasattr(process, 'proc'):  # Assume it is a Git.AutoInterrupt:
+            if isinstance(process, Git.AutoInterrupt):
                 process._terminate()
             else:  # Don't want to deal with the other case
-                raise RuntimeError(f"Thread join() timed out in cmd.handle_process_output()."
-                                   " kill_after_timeout={kill_after_timeout} seconds")
+                raise RuntimeError("Thread join() timed out in cmd.handle_process_output()."
+                                   f" kill_after_timeout={kill_after_timeout} seconds")
             if stderr_handler:
-                stderr_handler("error: process killed because it timed out."
-                               f" kill_after_timeout={kill_after_timeout} seconds")
+                error_str: Union[str, bytes] = (
+                    "error: process killed because it timed out."
+                    f" kill_after_timeout={kill_after_timeout} seconds")
+                if not decode_streams and isinstance(p_stderr, BinaryIO):
+                    #  Assume stderr_handler needs binary input
+                    error_str = cast(str, error_str)
+                    error_str = error_str.encode()
+                # We ignore typing on the next line because mypy does not like
+                # the way we infered that stderr takes str of bytes
+                stderr_handler(error_str)  # type: ignore
 
     if finalizer:
         return finalizer(process)
@@ -404,7 +412,7 @@ class Git(LazyMixin):
         def __init__(self, proc: Union[None, subprocess.Popen], args: Any) -> None:
             self.proc = proc
             self.args = args
-            self.status = None
+            self.status: Union[int, None] = None
 
         def _terminate(self) -> None:
             """Terminate the underlying process"""
@@ -447,8 +455,6 @@ class Git(LazyMixin):
                     call(("TASKKILL /F /T /PID %s 2>nul 1>nul" % str(proc.pid)), shell=True)
             # END exception handling
 
-
-
         def __del__(self) -> None:
             self._terminate()
 
@@ -465,11 +471,11 @@ class Git(LazyMixin):
             if stderr is None:
                 stderr_b = b''
             stderr_b = force_bytes(data=stderr, encoding='utf-8')
-
+            status: Union[int, None]
             if self.proc is not None:
                 status = self.proc.wait()
                 p_stderr = self.proc.stderr
-            else: #Assume the underlying proc was killed earlier or never existed
+            else:  # Assume the underlying proc was killed earlier or never existed
                 status = self.status
                 p_stderr = None
 

--- a/git/cmd.py
+++ b/git/cmd.py
@@ -79,7 +79,7 @@ def handle_process_output(process: 'Git.AutoInterrupt' | Popen,
                           finalizer: Union[None,
                                            Callable[[Union[subprocess.Popen, 'Git.AutoInterrupt']], None]] = None,
                           decode_streams: bool = True,
-                          timeout: float = 10.0) -> None:
+                          timeout: float = 60.0) -> None:
     """Registers for notifications to learn that process output is ready to read, and dispatches lines to
     the respective line handlers.
     This function returns once the finalizer returns

--- a/git/cmd.py
+++ b/git/cmd.py
@@ -94,7 +94,7 @@ def handle_process_output(process: 'Git.AutoInterrupt' | Popen,
         their contents to handlers.
         Set it to False if `universal_newline == True` (then streams are in text-mode)
         or if decoding must happen later (i.e. for Diffs).
-    :param timeout: float, timeout to pass to t.join() in case it hangs. Default = 10.0 seconds
+    :param timeout: float, or None timeout to pass to t.join() in case it hangs. Default = None.
     """
     # Use 2 "pump" threads and wait for both to finish.
     def pump_stream(cmdline: List[str], name: str, stream: Union[BinaryIO, TextIO], is_decode: bool,

--- a/git/cmd.py
+++ b/git/cmd.py
@@ -168,7 +168,7 @@ def handle_process_output(process: 'Git.AutoInterrupt' | Popen,
                     error_str = cast(str, error_str)
                     error_str = error_str.encode()
                 # We ignore typing on the next line because mypy does not like
-                # the way we infered that stderr takes str of bytes
+                # the way we inferred that stderr takes str or bytes
                 stderr_handler(error_str)  # type: ignore
 
     if finalizer:

--- a/git/remote.py
+++ b/git/remote.py
@@ -708,7 +708,7 @@ class Remote(LazyMixin, IterableObj):
 
     def _get_fetch_info_from_stderr(self, proc: 'Git.AutoInterrupt',
                                     progress: Union[Callable[..., Any], RemoteProgress, None],
-                                    timeout: Union[None, float] = None,
+                                    kill_after_timeout: Union[None, float] = None,
                                     ) -> IterableList['FetchInfo']:
 
         progress = to_progress_instance(progress)
@@ -726,7 +726,7 @@ class Remote(LazyMixin, IterableObj):
 
         progress_handler = progress.new_message_handler()
         handle_process_output(proc, None, progress_handler, finalizer=None, decode_streams=False,
-                              timeout=timeout)
+                              kill_after_timeout=kill_after_timeout)
 
         stderr_text = progress.error_lines and '\n'.join(progress.error_lines) or ''
         proc.wait(stderr=stderr_text)
@@ -772,7 +772,7 @@ class Remote(LazyMixin, IterableObj):
 
     def _get_push_info(self, proc: 'Git.AutoInterrupt',
                        progress: Union[Callable[..., Any], RemoteProgress, None],
-                       timeout: Union[None, float] = None) -> IterableList[PushInfo]:
+                       kill_after_timeout: Union[None, float] = None) -> IterableList[PushInfo]:
         progress = to_progress_instance(progress)
 
         # read progress information from stderr
@@ -790,7 +790,7 @@ class Remote(LazyMixin, IterableObj):
                 pass
 
         handle_process_output(proc, stdout_handler, progress_handler, finalizer=None, decode_streams=False,
-                              timeout=timeout)
+                              kill_after_timeout=kill_after_timeout)
         stderr_text = progress.error_lines and '\n'.join(progress.error_lines) or ''
         try:
             proc.wait(stderr=stderr_text)
@@ -817,7 +817,8 @@ class Remote(LazyMixin, IterableObj):
 
     def fetch(self, refspec: Union[str, List[str], None] = None,
               progress: Union[RemoteProgress, None, 'UpdateProgress'] = None,
-              verbose: bool = True, timeout: Union[None, float] = None,
+              verbose: bool = True,
+              kill_after_timeout: Union[None, float] = None,
               **kwargs: Any) -> IterableList[FetchInfo]:
         """Fetch the latest changes for this remote
 
@@ -838,6 +839,9 @@ class Remote(LazyMixin, IterableObj):
             for 'refspec' will make use of this facility.
         :param progress: See 'push' method
         :param verbose: Boolean for verbose output
+        :param kill_after_timeout:
+            To specify a timeout in seconds for the git command, after which the process
+            should be killed. It is set to None by default.
         :param kwargs: Additional arguments to be passed to git-fetch
         :return:
             IterableList(FetchInfo, ...) list of FetchInfo instances providing detailed
@@ -858,20 +862,22 @@ class Remote(LazyMixin, IterableObj):
 
         proc = self.repo.git.fetch(self, *args, as_process=True, with_stdout=False,
                                    universal_newlines=True, v=verbose, **kwargs)
-        res = self._get_fetch_info_from_stderr(proc, progress, timeout=timeout)
+        res = self._get_fetch_info_from_stderr(proc, progress,
+                                               kill_after_timeout=kill_after_timeout)
         if hasattr(self.repo.odb, 'update_cache'):
             self.repo.odb.update_cache()
         return res
 
     def pull(self, refspec: Union[str, List[str], None] = None,
              progress: Union[RemoteProgress, 'UpdateProgress', None] = None,
-             timeout: Union[None, float] = None,
+             kill_after_timeout: Union[None, float] = None,
              **kwargs: Any) -> IterableList[FetchInfo]:
         """Pull changes from the given branch, being the same as a fetch followed
         by a merge of branch with your local branch.
 
         :param refspec: see 'fetch' method
         :param progress: see 'push' method
+        :param kill_after_timeout: see 'fetch' method
         :param kwargs: Additional arguments to be passed to git-pull
         :return: Please see 'fetch' method """
         if refspec is None:
@@ -880,14 +886,16 @@ class Remote(LazyMixin, IterableObj):
         kwargs = add_progress(kwargs, self.repo.git, progress)
         proc = self.repo.git.pull(self, refspec, with_stdout=False, as_process=True,
                                   universal_newlines=True, v=True, **kwargs)
-        res = self._get_fetch_info_from_stderr(proc, progress, timeout=timeout)
+        res = self._get_fetch_info_from_stderr(proc, progress,
+                                               kill_after_timeout=kill_after_timeout)
         if hasattr(self.repo.odb, 'update_cache'):
             self.repo.odb.update_cache()
         return res
 
     def push(self, refspec: Union[str, List[str], None] = None,
              progress: Union[RemoteProgress, 'UpdateProgress', Callable[..., RemoteProgress], None] = None,
-             timeout: Union[None, float] = None, **kwargs: Any) -> IterableList[PushInfo]:
+             kill_after_timeout: Union[None, float] = None,
+             **kwargs: Any) -> IterableList[PushInfo]:
         """Push changes from source branch in refspec to target branch in refspec.
 
         :param refspec: see 'fetch' method
@@ -903,6 +911,9 @@ class Remote(LazyMixin, IterableObj):
               overrides the ``update()`` function.
 
         :note: No further progress information is returned after push returns.
+        :param kill_after_timeout:
+            To specify a timeout in seconds for the git command, after which the process
+            should be killed. It is set to None by default.
         :param kwargs: Additional arguments to be passed to git-push
         :return:
             list(PushInfo, ...) list of PushInfo instances, each
@@ -914,8 +925,11 @@ class Remote(LazyMixin, IterableObj):
             be 0."""
         kwargs = add_progress(kwargs, self.repo.git, progress)
         proc = self.repo.git.push(self, refspec, porcelain=True, as_process=True,
-                                  universal_newlines=True, **kwargs)
-        return self._get_push_info(proc, progress, timeout=timeout)
+                                  universal_newlines=True,
+                                  kill_after_timeout=kill_after_timeout,
+                                  **kwargs)
+        return self._get_push_info(proc, progress,
+                                   kill_after_timeout=kill_after_timeout)
 
     @ property
     def config_reader(self) -> SectionConstraint[GitConfigParser]:

--- a/git/remote.py
+++ b/git/remote.py
@@ -795,6 +795,8 @@ class Remote(LazyMixin, IterableObj):
         try:
             proc.wait(stderr=stderr_text)
         except Exception:
+            # This is different than fetch (which fails if there is any std_err
+            # even if there is an output)
             if not output:
                 raise
             elif stderr_text:

--- a/git/remote.py
+++ b/git/remote.py
@@ -707,7 +707,8 @@ class Remote(LazyMixin, IterableObj):
         return self
 
     def _get_fetch_info_from_stderr(self, proc: 'Git.AutoInterrupt',
-                                    progress: Union[Callable[..., Any], RemoteProgress, None]
+                                    progress: Union[Callable[..., Any], RemoteProgress, None],
+                                    timeout: float = 60.0
                                     ) -> IterableList['FetchInfo']:
 
         progress = to_progress_instance(progress)
@@ -724,7 +725,8 @@ class Remote(LazyMixin, IterableObj):
         cmds = set(FetchInfo._flag_map.keys())
 
         progress_handler = progress.new_message_handler()
-        handle_process_output(proc, None, progress_handler, finalizer=None, decode_streams=False)
+        handle_process_output(proc, None, progress_handler, finalizer=None, decode_streams=False,
+                              timeout=timeout)
 
         stderr_text = progress.error_lines and '\n'.join(progress.error_lines) or ''
         proc.wait(stderr=stderr_text)
@@ -769,7 +771,8 @@ class Remote(LazyMixin, IterableObj):
         return output
 
     def _get_push_info(self, proc: 'Git.AutoInterrupt',
-                       progress: Union[Callable[..., Any], RemoteProgress, None]) -> IterableList[PushInfo]:
+                       progress: Union[Callable[..., Any], RemoteProgress, None],
+                       timeout: float = 60.0) -> IterableList[PushInfo]:
         progress = to_progress_instance(progress)
 
         # read progress information from stderr
@@ -786,7 +789,8 @@ class Remote(LazyMixin, IterableObj):
                 # If an error happens, additional info is given which we parse below.
                 pass
 
-        handle_process_output(proc, stdout_handler, progress_handler, finalizer=None, decode_streams=False)
+        handle_process_output(proc, stdout_handler, progress_handler, finalizer=None, decode_streams=False,
+                              timeout=timeout)
         stderr_text = progress.error_lines and '\n'.join(progress.error_lines) or ''
         try:
             proc.wait(stderr=stderr_text)
@@ -813,7 +817,8 @@ class Remote(LazyMixin, IterableObj):
 
     def fetch(self, refspec: Union[str, List[str], None] = None,
               progress: Union[RemoteProgress, None, 'UpdateProgress'] = None,
-              verbose: bool = True, **kwargs: Any) -> IterableList[FetchInfo]:
+              verbose: bool = True, timeout: float = 60.0,
+              **kwargs: Any) -> IterableList[FetchInfo]:
         """Fetch the latest changes for this remote
 
         :param refspec:
@@ -853,13 +858,14 @@ class Remote(LazyMixin, IterableObj):
 
         proc = self.repo.git.fetch(self, *args, as_process=True, with_stdout=False,
                                    universal_newlines=True, v=verbose, **kwargs)
-        res = self._get_fetch_info_from_stderr(proc, progress)
+        res = self._get_fetch_info_from_stderr(proc, progress, timeout=timeout)
         if hasattr(self.repo.odb, 'update_cache'):
             self.repo.odb.update_cache()
         return res
 
     def pull(self, refspec: Union[str, List[str], None] = None,
              progress: Union[RemoteProgress, 'UpdateProgress', None] = None,
+             timeout: float = 60.0,
              **kwargs: Any) -> IterableList[FetchInfo]:
         """Pull changes from the given branch, being the same as a fetch followed
         by a merge of branch with your local branch.
@@ -874,14 +880,14 @@ class Remote(LazyMixin, IterableObj):
         kwargs = add_progress(kwargs, self.repo.git, progress)
         proc = self.repo.git.pull(self, refspec, with_stdout=False, as_process=True,
                                   universal_newlines=True, v=True, **kwargs)
-        res = self._get_fetch_info_from_stderr(proc, progress)
+        res = self._get_fetch_info_from_stderr(proc, progress, timeout=timeout)
         if hasattr(self.repo.odb, 'update_cache'):
             self.repo.odb.update_cache()
         return res
 
     def push(self, refspec: Union[str, List[str], None] = None,
              progress: Union[RemoteProgress, 'UpdateProgress', Callable[..., RemoteProgress], None] = None,
-             **kwargs: Any) -> IterableList[PushInfo]:
+             timeout: float = 60.0, **kwargs: Any) -> IterableList[PushInfo]:
         """Push changes from source branch in refspec to target branch in refspec.
 
         :param refspec: see 'fetch' method
@@ -909,7 +915,7 @@ class Remote(LazyMixin, IterableObj):
         kwargs = add_progress(kwargs, self.repo.git, progress)
         proc = self.repo.git.push(self, refspec, porcelain=True, as_process=True,
                                   universal_newlines=True, **kwargs)
-        return self._get_push_info(proc, progress)
+        return self._get_push_info(proc, progress, timeout=timeout)
 
     @ property
     def config_reader(self) -> SectionConstraint[GitConfigParser]:

--- a/git/remote.py
+++ b/git/remote.py
@@ -708,7 +708,7 @@ class Remote(LazyMixin, IterableObj):
 
     def _get_fetch_info_from_stderr(self, proc: 'Git.AutoInterrupt',
                                     progress: Union[Callable[..., Any], RemoteProgress, None],
-                                    timeout: float = 60.0
+                                    timeout: Union[None, float] = None,
                                     ) -> IterableList['FetchInfo']:
 
         progress = to_progress_instance(progress)
@@ -772,7 +772,7 @@ class Remote(LazyMixin, IterableObj):
 
     def _get_push_info(self, proc: 'Git.AutoInterrupt',
                        progress: Union[Callable[..., Any], RemoteProgress, None],
-                       timeout: float = 60.0) -> IterableList[PushInfo]:
+                       timeout: Union[None, float] = None) -> IterableList[PushInfo]:
         progress = to_progress_instance(progress)
 
         # read progress information from stderr
@@ -817,7 +817,7 @@ class Remote(LazyMixin, IterableObj):
 
     def fetch(self, refspec: Union[str, List[str], None] = None,
               progress: Union[RemoteProgress, None, 'UpdateProgress'] = None,
-              verbose: bool = True, timeout: float = 60.0,
+              verbose: bool = True, timeout: Union[None, float] = None,
               **kwargs: Any) -> IterableList[FetchInfo]:
         """Fetch the latest changes for this remote
 
@@ -865,7 +865,7 @@ class Remote(LazyMixin, IterableObj):
 
     def pull(self, refspec: Union[str, List[str], None] = None,
              progress: Union[RemoteProgress, 'UpdateProgress', None] = None,
-             timeout: float = 60.0,
+             timeout: Union[None, float] = None,
              **kwargs: Any) -> IterableList[FetchInfo]:
         """Pull changes from the given branch, being the same as a fetch followed
         by a merge of branch with your local branch.
@@ -887,7 +887,7 @@ class Remote(LazyMixin, IterableObj):
 
     def push(self, refspec: Union[str, List[str], None] = None,
              progress: Union[RemoteProgress, 'UpdateProgress', Callable[..., RemoteProgress], None] = None,
-             timeout: float = 60.0, **kwargs: Any) -> IterableList[PushInfo]:
+             timeout: Union[None, float] = None, **kwargs: Any) -> IterableList[PushInfo]:
         """Push changes from source branch in refspec to target branch in refspec.
 
         :param refspec: see 'fetch' method

--- a/test/test_remote.py
+++ b/test/test_remote.py
@@ -658,7 +658,7 @@ class TestRemote(TestBase):
 class TestTimeouts(TestBase):
     @with_rw_repo('HEAD', bare=False)
     def test_timeout_funcs(self, repo):
-        for function in ["pull"]: #"can't get fetch and push to reliably timeout
+        for function in ["pull"]:  # can't get fetch and push to reliably timeout
             f = getattr(repo.remotes.origin, function)
             assert f is not None  # Make sure these functions exist
             _ = f() # Make sure the function runs

--- a/test/test_remote.py
+++ b/test/test_remote.py
@@ -406,7 +406,7 @@ class TestRemote(TestBase):
         # cleanup - delete created tags and branches as we are in an innerloop on
         # the same repository
         TagReference.delete(rw_repo, new_tag, other_tag)
-        remote.push(":%s" % other_tag.path)
+        remote.push(":%s" % other_tag.path, timeout=10.0)
 
     @skipIf(HIDE_WINDOWS_FREEZE_ERRORS, "FIXME: Freezes!")
     @with_rw_and_rw_remote_repo('0.1.6')

--- a/test/test_remote.py
+++ b/test/test_remote.py
@@ -6,6 +6,7 @@
 
 import random
 import tempfile
+import pytest
 from unittest import skipIf
 
 from git import (
@@ -401,12 +402,12 @@ class TestRemote(TestBase):
         res = remote.push(all=True)
         self._do_test_push_result(res, remote)
 
-        remote.pull('master', timeout=10.0)
+        remote.pull('master', kill_after_timeout=10.0)
 
         # cleanup - delete created tags and branches as we are in an innerloop on
         # the same repository
         TagReference.delete(rw_repo, new_tag, other_tag)
-        remote.push(":%s" % other_tag.path, timeout=10.0)
+        remote.push(":%s" % other_tag.path, kill_after_timeout=10.0)
 
     @skipIf(HIDE_WINDOWS_FREEZE_ERRORS, "FIXME: Freezes!")
     @with_rw_and_rw_remote_repo('0.1.6')
@@ -467,7 +468,8 @@ class TestRemote(TestBase):
             # Only for remotes - local cases are the same or less complicated
             # as additional progress information will never be emitted
             if remote.name == "daemon_origin":
-                self._do_test_fetch(remote, rw_repo, remote_repo, timeout=10.0)
+                self._do_test_fetch(remote, rw_repo, remote_repo,
+                                    kill_after_timeout=10.0)
                 ran_fetch_test = True
             # END fetch test
 
@@ -651,3 +653,15 @@ class TestRemote(TestBase):
         rem = repo.remote('origin')
         with self.assertRaisesRegex(GitCommandError, "src refspec __BAD_REF__ does not match any"):
             rem.push('__BAD_REF__')
+
+
+class TestTimeouts(TestBase):
+    @with_rw_repo('HEAD', bare=False)
+    def test_timeout_funcs(self, repo):
+        for function in ["pull", "fetch"]: #"can't get push to reliably timeout
+            f = getattr(repo.remotes.origin, function)
+            assert f is not None  # Make sure these functions exist
+
+            with self.assertRaisesRegex(GitCommandError,
+                                        "kill_after_timeout=0.01 s"):
+                f(kill_after_timeout=0.01)

--- a/test/test_remote.py
+++ b/test/test_remote.py
@@ -658,10 +658,16 @@ class TestRemote(TestBase):
 class TestTimeouts(TestBase):
     @with_rw_repo('HEAD', bare=False)
     def test_timeout_funcs(self, repo):
-        for function in ["pull"]:  # can't get fetch and push to reliably timeout
+        # Force error code to prevent a race condition if the python thread is
+        # slow
+        default = Git.AutoInterrupt._status_code_if_terminate
+        Git.AutoInterrupt._status_code_if_terminate = -15
+        for function in ["pull", "fetch"]:  # can't get push to timeout
             f = getattr(repo.remotes.origin, function)
             assert f is not None  # Make sure these functions exist
-            _ = f() # Make sure the function runs
+            _ = f()  # Make sure the function runs
             with pytest.raises(GitCommandError,
-                               match="kill_after_timeout=0.001 s"):
-                f(kill_after_timeout=0.001)
+                               match="kill_after_timeout=0 s"):
+                f(kill_after_timeout=0)
+
+        Git.AutoInterrupt._status_code_if_terminate = default

--- a/test/test_remote.py
+++ b/test/test_remote.py
@@ -401,7 +401,7 @@ class TestRemote(TestBase):
         res = remote.push(all=True)
         self._do_test_push_result(res, remote)
 
-        remote.pull('master')
+        remote.pull('master', timeout=10.0)
 
         # cleanup - delete created tags and branches as we are in an innerloop on
         # the same repository
@@ -467,7 +467,7 @@ class TestRemote(TestBase):
             # Only for remotes - local cases are the same or less complicated
             # as additional progress information will never be emitted
             if remote.name == "daemon_origin":
-                self._do_test_fetch(remote, rw_repo, remote_repo)
+                self._do_test_fetch(remote, rw_repo, remote_repo, timeout=10.0)
                 ran_fetch_test = True
             # END fetch test
 

--- a/test/test_remote.py
+++ b/test/test_remote.py
@@ -663,5 +663,5 @@ class TestTimeouts(TestBase):
             assert f is not None  # Make sure these functions exist
             _ = f() # Make sure the function runs
             with pytest.raises(GitCommandError,
-                               match="kill_after_timeout=0 s"):
-                f(kill_after_timeout=0)
+                               match="kill_after_timeout=0.001 s"):
+                f(kill_after_timeout=0.001)

--- a/test/test_remote.py
+++ b/test/test_remote.py
@@ -661,7 +661,7 @@ class TestTimeouts(TestBase):
         for function in ["pull", "fetch"]: #"can't get push to reliably timeout
             f = getattr(repo.remotes.origin, function)
             assert f is not None  # Make sure these functions exist
-
-            with self.assertRaisesRegex(GitCommandError,
-                                        "kill_after_timeout=0.01 s"):
+            _ = f() # Make sure the function runs
+            with pytest.raises(GitCommandError,
+                               match="kill_after_timeout=0.01 s"):
                 f(kill_after_timeout=0.01)

--- a/test/test_remote.py
+++ b/test/test_remote.py
@@ -663,5 +663,5 @@ class TestTimeouts(TestBase):
             assert f is not None  # Make sure these functions exist
             _ = f() # Make sure the function runs
             with pytest.raises(GitCommandError,
-                               match="kill_after_timeout=0.01 s"):
-                f(kill_after_timeout=0.01)
+                               match="kill_after_timeout=0 s"):
+                f(kill_after_timeout=0)

--- a/test/test_remote.py
+++ b/test/test_remote.py
@@ -658,7 +658,7 @@ class TestRemote(TestBase):
 class TestTimeouts(TestBase):
     @with_rw_repo('HEAD', bare=False)
     def test_timeout_funcs(self, repo):
-        for function in ["pull", "fetch"]: #"can't get push to reliably timeout
+        for function in ["pull"]: #"can't get fetch and push to reliably timeout
             f = getattr(repo.remotes.origin, function)
             assert f is not None  # Make sure these functions exist
             _ = f() # Make sure the function runs

--- a/test/test_remote.py
+++ b/test/test_remote.py
@@ -164,7 +164,7 @@ class TestRemote(TestBase):
         index.commit("Committing %s" % new_file)
         return new_file
 
-    def _do_test_fetch(self, remote, rw_repo, remote_repo):
+    def _do_test_fetch(self, remote, rw_repo, remote_repo, **kwargs):
         # specialized fetch testing to de-clutter the main test
         self._do_test_fetch_info(rw_repo)
 
@@ -183,7 +183,7 @@ class TestRemote(TestBase):
         # put remote head to master as it is guaranteed to exist
         remote_repo.head.reference = remote_repo.heads.master
 
-        res = fetch_and_test(remote)
+        res = fetch_and_test(remote, **kwargs)
         # all up to date
         for info in res:
             self.assertTrue(info.flags & info.HEAD_UPTODATE)

